### ISSUE TITLE
Popping up cider-error buffer for exceptions is slow

### ISSUE THREAD

**Issue by alexander-yakushev** [clojure-emacs/cider#2237]
I'm not sure when did that start, but for quite a while now, whenever an exception happens that propagates all the way up to `*cider-error*` buffer, it takes a noticeable amount of time (~0.5-1 second) for the buffer to show up. This is reproducible every time and on two different machines/configs.

This doesn't seem to be the loading of middleware (happens every time, not only the first time), nor it's the cost of printing the stacktrace (happens instantly if I print the Exception object to REPL).

Sorry if this is reported already, couldn't find anything similar.

## Expected behavior

Exception buffer should appear instantly when an exception is thrown.

## Actual behavior

CIDER lags for 0.5-1 second before showing the error buffer.

## Steps to reproduce the problem

`(/ 1 0)` either in the REPL or the buffer, with `cider-show-error-buffer t`.

## Environment & Version information

MacOS, Emacs 25.3.1, cider-nrepl 0.16.0, CIDER 20180217.336, boot 2.7.2.

```
;; Connected to nREPL server - nrepl://localhost:50285
;; CIDER 0.17.0snapshot (package: 20180217.336), nREPL 0.2.13
;; Clojure 1.9.0, Java 1.8.0_102
```

**Comment by bbatsov** [comment]
Can you attach a bit of profiling data to the ticket?

**Comment by alexander-yakushev** [comment]
I tried profiling it, but since it is asynchronous, I don't know how to schedule it, and if I start/stop manually, lot's of irrelevant stuff goes into the profile.

**Comment by alexander-yakushev** [comment]
Something like this (obtained through standing in the buffer or `(/ 1 0)` and doing 
`M-: (progn (profiler-start 'cpu) (cider-eval-last-sexp) (sleep-for 1) (profiler-report) (profiler-stop))`

[profiler-profile "24.3" cpu #s(hash-table size 65 test equal rehash-size 1.5 rehash-threshold 0.8 data ([profiler-calltree-build-unified profiler-calltree-build profiler-report-render-calltree-1 profiler-report-rerender-calltree profiler-report-render-calltree profiler-report-setup-buffer profiler-report-profile-other-window profiler-report-cpu profiler-report progn eval eval-expression funcall-interactively "#<subr call-interactively>" ad-Advice-call-interactively apply] 1 [ad-Advice-select-window apply select-window pop-to-buffer switch-to-buffer-other-window profiler-report-profile-other-window profiler-report-cpu profiler-report progn eval eval-expression funcall-interactively "#<subr call-interactively>" ad-Advice-call-interactively apply call-interactively] 1 [nrepl-make-response-handler cider-interactive-eval-handler "#<compiled 0x43ce046b>" cider-map-connections cider-interactive-eval cider-eval-last-sexp progn eval eval-expression funcall-interactively "#<subr call-interactively>" ad-Advice-call-interactively apply call-interactively command-execute nil] 1 [pp nrepl-log-pp-object nrepl-log--pp-listlike nrepl-log-pp-object nrepl-log-message nrepl-send-request nrepl-request:eval cider-nrepl-request:eval "#<compiled 0x43ce046b>" cider-map-connections cider-interactive-eval cider-eval-last-sexp progn eval eval-expression funcall-interactively] 1 [replace-regexp-in-string nrepl--bdecode-1 nrepl--bdecode-list nrepl--bdecode-1 nrepl--bdecode-message nrepl-bdecode nrepl-client-filter sleep-for progn eval eval-expression funcall-interactively "#<subr call-interactively>" ad-Advice-call-interactively apply call-interactively] 1 [cider-connections cider-current-connection cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter sleep-for progn eval eval-expression funcall-interactively "#<subr call-interactively>" ad-Advice-call-interactively] 1 [nrepl-bencode mapconcat nrepl-bencode nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter sleep-for progn eval eval-expression funcall-interactively] 15 [nrepl-bencode nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter sleep-for progn eval eval-expression funcall-interactively "#<subr call-interactively>" ad-Advice-call-interactively] 1 [nrepl-connection-identifier nrepl-messages-buffer-name nrepl-messages-buffer nrepl-log-message nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter sleep-for progn eval eval-expression] 1 [apply seq-drop setq while let if progn seq-partition nrepl-log--pp-listlike nrepl-log-pp-object nrepl-log-message nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>"] 1 [cons setq while let if progn seq-partition nrepl-log--pp-listlike nrepl-log-pp-object nrepl-log-message nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>"] 1 [setq while let progn "#<lambda 0x27cc74d57dbaf782>" apply seq-take cons setq while let if progn seq-partition nrepl-log--pp-listlike nrepl-log-pp-object] 1 [progn seq-max nrepl-log--pp-listlike nrepl-log-pp-object nrepl-log-message nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter sleep-for progn eval] 1 ["#<lambda 0x6bfded1880138a8>" mapc progn seq-do catch progn seq-some progn seq-contains "#<compiled 0x414800b3>" funcall not "#<lambda 0x1123402d21a64053>" funcall if "#<lambda 0x112bb0f15b344c00>"] 1 [nrepl-log-pp-object nrepl-log-message nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter sleep-for progn eval eval-expression funcall-interactively "#<subr call-interactively>"] 5 [down-list pp-buffer pp-to-string pp nrepl-log-pp-object nrepl-log--pp-listlike nrepl-log-pp-object nrepl-log-message nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter] 1 [pp-buffer pp-to-string pp nrepl-log-pp-object nrepl-log--pp-listlike nrepl-log-pp-object nrepl-log-message nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter sleep-for] 1 [generate-new-buffer pp-to-string pp nrepl-log-pp-object nrepl-log--pp-listlike nrepl-log-pp-object nrepl-log-message nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter sleep-for] 20 [pp nrepl-log-pp-object nrepl-log--pp-listlike nrepl-log-pp-object nrepl-log-message nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter sleep-for progn eval] 3 [nrepl-log--pp-listlike nrepl-log-pp-object nrepl-log-message nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter sleep-for progn eval eval-expression funcall-interactively] 1 [pp-to-string pp nrepl-log-pp-object nrepl-log--pp-listlike nrepl-log-pp-object nrepl-log-message nrepl-send-request cider-nrepl-send-request cider-default-err-op-handler cider-default-err-handler "#<compiled 0x43c9a0c5>" "#<compiled 0x4420c3ef>" nrepl--dispatch-response nrepl-client-filter sleep-for progn] 2 [nrepl-log--pp-listlike nrepl-log-pp-object nrepl-log-message nrepl--dispatch-response nrepl-client-filter sleep-for progn eval eval-expression funcall-interactively "#<subr call-interactively>" ad-Advice-call-interactively apply call-interactively command-execute nil] 1 [progn eval eval-expression funcall-interactively "#<subr call-interactively>" ad-Advice-call-interactively apply call-interactively command-execute nil nil nil nil nil nil nil] 4 [nrepl-log-pp-object nrepl-log-message nrepl--dispatch-response nrepl-client-filter sleep-for progn eval eval-expression funcall-interactively "#<subr call-interactively>" ad-Advice-call-interactively apply call-interactively command-execute nil nil] 1 [cider-make-popup-buffer cider-popup-buffer cider-new-error-buffer cider--render-stacktrace-causes cider--handle-stacktrace-response "#<compiled 0x41545d6f>" nrepl--dispatch-response nrepl-client-filter sleep-for progn eval eval-expression funcall-interactively "#<subr call-interactively>" ad-Advice-call-interactively apply] 1 [cons "#<lambda 0xbb3e3b36>" mapcar fci-get-visible-ranges let fci-delete-unneeded progn if save-current-buffer while let let* fci-redraw-frame set-window-buffer window--display-buffer display-buffer-in-previous-window] 3 [fci-redraw-region fci-redraw-window while let fci-update-all-windows progn if save-current-buffer while let let* fci-redraw-frame set-window-buffer window--display-buffer display-buffer-in-previous-window display-buffer] 1 [forward-paragraph fill-forward-paragraph fill-region cider-stacktrace-emit-indented cider-stacktrace-render-cause cider-stacktrace-render cider--render-stacktrace-causes cider--handle-stacktrace-response "#<compiled 0x41545d6f>" nrepl--dispatch-response nrepl-client-filter sleep-for progn eval eval-expression funcall-interactively] 1 [cider-stacktrace-indicate-filters cider-stacktrace-apply-filters cider-stacktrace-apply-cause-visibility cider-stacktrace-cycle-cause cider-stacktrace-initialize cider-stacktrace-render cider--render-stacktrace-causes cider--handle-stacktrace-response "#<compiled 0x41545d6f>" nrepl--dispatch-response nrepl-client-filter sleep-for progn eval eval-expression funcall-interactively] 1 [profiler-cpu-profile profiler-report-cpu profiler-report progn eval eval-expression funcall-interactively "#<subr call-interactively>" ad-Advice-call-interactively apply call-interactively command-execute nil nil nil nil] 5 [Automatic\ GC] 207)) (23207 41682 135288 0) nil]

**Comment by xiongtx** [comment]
It seems you're spending a lot of time on GC:

![image](https://user-images.githubusercontent.com/1624090/37386857-731838aa-2717-11e8-90ee-59433c2174f5.png)

What's your memory usage look like when displaying these buffers?

I couldn't replicate this issue with `CIDER 0.17.0snapshot (package: 20180312.26)`. Can you try upgrading to see if the problem persists?

**Comment by bbatsov** [comment]
I guess another thing to try would be increasing the GC threshold, so that the GC is kicking in less often. Or reduce it if you made it huge in the past (see http://bling.github.io/blog/2016/01/18/why-are-you-changing-gc-cons-threshold/)

**Comment by xiongtx** [comment]
Unless GC threshold had been set much lower than normal, it shouldn't be a problem. Something else is likely going on.

**Comment by alexander-yakushev** [comment]
OK, I spent almost a day fiddling with my Emacs config. Good news: I've updated all the packages (including CIDER) and this exact issue is gone for me. Whew!

I cannot verify now whether the problem was with older CIDER package or my overall Emacs setup. Much likely the latter. I still have lags when pressing Smex' `M-x` and they also seem GC-related. Raising GC threshold seems to help only for a few minutes, then it's bad again.

In any case, the tip for similar misfortuners as me: update everything and check if your overall Emacs installation is not bogged down by something.

Thanks for help!
